runtime/src/main/java/com/dylibso/chicory/runtime/InterpreterMachine.java

```diff
@@ -2286,56 +2272,84 @@ private static void I64_ATOMIC_STORE(MStack stack, Instance instance, Operands o
         if (ptr % 8 != 0) {
             throw new InvalidException("unaligned atomic");
         }
-        synchronized (instance.memory().lock(ptr)) {
-            instance.memory().writeLong(ptr, value);
-        }
+        instance.memory().atomicWriteLong(ptr, value);
     }
 
     private static void I32_ATOMIC_RMW(
-            MStack stack,
-            Instance instance,
-            Operands operands,
-            BiFunction<Integer, Integer, Integer> op) {
-        var toOp = (int) stack.pop();
+            MStack stack, Instance instance, Operands operands, AtomicOp op) {
+        var operand = (int) stack.pop();
         var ptr = readMemPtr(stack, operands);
         if (ptr % 4 != 0) {
             throw new InvalidException("unaligned atomic");
         }
-        synchronized (instance.memory().lock(ptr)) {
-            var val = (int) instance.memory().readI32(ptr);
-            instance.memory().writeI32(ptr, op.apply(val, toOp));
-            stack.push(val);
-        }
+        int oldVal;
+        switch (op) {
+            case ADD:
+                oldVal = instance.memory().atomicAddInt(ptr, operand);
+                break;
+            case SUB:
+                oldVal = instance.memory().atomicAddInt(ptr, -operand);
+                break;
+            case AND:
+                oldVal = instance.memory().atomicAndInt(ptr, operand);
+                break;
+            case OR:
+                oldVal = instance.memory().atomicOrInt(ptr, operand);
+                break;
+            case XOR:
+                oldVal = instance.memory().atomicXorInt(ptr, operand);
+                break;
+            case XCHG:
+                oldVal = instance.memory().atomicXchgInt(ptr, operand);
+                break;
+            default:
+                throw new IllegalStateException("Unexpected atomic op: " + op);
+        }
+        stack.push(oldVal);
     }
 
     private static void I32_ATOMIC_RMW_CMPXCHG(MStack stack, Instance instance, Operands operands) {
-        var replacement = (int) stack.pop(); // c3
-        var expected = (int) stack.pop(); // c2
-        var ptr = readMemPtr(stack, operands); // i
+        var replacement = (int) stack.pop();
```

Comment:
nit: I think the names in the comments are reflecting the spec, I'd leave it for the future reader.